src/control/cmd/daos/README.md

```diff
@@ -0,0 +1,155 @@
+
+## Implementation Details
+
+This version of the daos utility is implemented using a hybrid approach where the user interface is written in Go and uses the <a href="https://github.com/jessevdk/go-flags">go-flags</a> argument-parsing library to handle processing of command-line arguments and other UI-related tasks. The frontend code then uses <a href="https://golang.org/cmd/cgo/">cgo</a> to interface with the C-based backend handlers.
+
+The original version of this utility was implemented entirely in C. There were advantages to this approach, particularly that the language is familiar to developers who primarily work on the DAOS client and server code. The pure-C implementation does suffer from some problems, however. Notably, the UI is very different from other user-facing utilities (dmg, etc.), and adding usability-improving features (tab completion, etc.) is much more difficult in C than it is in Go.
+
+## Layout/Organization
+
+The frontend code under src/control/cmd/daos is organized along the following lines:
+  * main.go: The entry point to the program. Handles parsing of arguments and delegating to subcommand handlers. Most feature development efforts should not need to make any modifications to this file.
+  * container.go: This is where most of the container subcommand (`daos cont ...`) handlers are. The top-level `containerCmd` struct defines handlers for each of the subcommands.
+  * filesystem.go: This file contains the handlers for the `daos fs ...` family of commands. The top-level `fsCmd` struct defines handlers for each of the subcommands.
+  * pool.go: This file contains the handlers for the `daos pool` family of commands. The top-level `poolCmd` struct defines handlers for each of the subcommands.
+
+The remainder of the files contain shared utility code or have broken out some of the container subcommand handling in order to reduce the size of container.go.
+
+The backend code is entirely written in C and for the moment is organized in the src/utils directory. The following files are linked into the `daos_cmd_hdlrs` library so that they can be called from the Go frontend:
+  * daos_autotest.c
+  * daos_dfs_hdlr.c
+  * daos_hdlr.c
+
+Note that the daos.c file contains frontend code for the C-only implementation, and its contents are not available to be utilized from Go.
+
+Longer-term, it would probably make sense to move the source files for `daos_cmd_hdlrs` under src/control/cmd/daos so that they live alongside the frontend code and can be compiled directly into the new daos binary in order to remove the need for the shared library. We could also consider implementing more of the backend logic in Go, but this may not necessarily make sense in many cases (e.g. data mover).
```

Comment:
If the shared library is only for the linking across directories then we can probably do that better.  I think what you need to do is define something in the top-level SConstruct, export it then add to it in one subdir and use it in another.  You can't export from one sub-dir to another though.  @jolivier23 probably has further thoughts here.